### PR TITLE
ErrorReportWindow: add model and drop 'title'

### PR DESCRIPTION
**Pull request by DanVanAtta**
## Overview

- Remove user input form field 'subject' from error report window

- Add a second auto-filled form field for 'additional informaton' on the
report window. This second text area is pre-filled with any error
message or contents from the console. If neither of those, then the text
area is not rendered at all. This second text field is write-enabled to
allow users to update contents if desired.

- Enhancement: Error report window 'submit' button is enabled only when
a minimum amount of text is present.

- Adds a model object to remove logic burden from window class

### Before
screenshot from 2019-01-09 01-26-23

### After

Example without 'error info':
screenshot from 2019-01-09 01-19-46

Second example with 'error info':
screenshot from 2019-01-09 01-22-41

### CODE REVIEW

File: game-core/src/main/java/swinglib/JTextAreaBuilder.java [hunk 1]
```diff
@@ -38,6 +40,32 @@ public static JTextAreaBuilder builder() {
...
   }
 
+  /**
+   * Attaches a listener that is fired whenever text is typed in the textArea.
+   * Note, the listener will be notified if text is either typed or pasted.
+   *
+   * @param textArea The component to which we attach a listener.
+   * @param listener The listener that should be invoked on key events.
+   */
+  public static void addTextListener(final JTextArea textArea, final Runnable listener) {
```
Comment (ssoloff) [inline game-core/src/main/java/swinglib/JTextAreaBuilder.java]: Does this need to be replaced in favor of the new `DocumentListenerBuilder#attachDocumentListener()` method?

File: game-core/src/main/java/games/strategy/debug/error/reporting/ErrorReportWindow.java [hunk 2]
```diff
@@ -1,82 +1,141 @@
...
+        .columns(10)
+        .rows(2)
         .build();
 
+    final JTextArea additionalInfo = model.getAttachedData().map(
+        attached -> JTextAreaBuilder.builder()
+            .rows(5)
+            .text(attached)
+            .build())
+        .orElse(new JTextArea());
```
[the remark this answers is not in the document]
Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/debug/error/reporting/ErrorReportWindow.java]: Good suggestion, but there is a deeper fix, a dry violation of when we have a text area that should be fixed first.

This UI code I'm finding is a bit hard to follow as well, I'm working on both, stay tuned.

File: game-core/src/main/java/swinglib/JTextAreaBuilder.java (hunk 1 above)
Comment (DanVanAtta) [reply game-core/src/main/java/swinglib/JTextAreaBuilder.java]: Yup, `DocumentListenerBuilder` was opened while this PR was in flight, nonetheless it is a good cleanup, good eye.

Done with these two updates:
- Call to this method removed in:  2c252c2 
- Code removed in: 9b5736e

File: game-core/src/main/java/games/strategy/debug/error/reporting/ErrorReportWindow.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/debug/error/reporting/ErrorReportWindow.java]: b3ebb74 has updates that update this.

I was not very happy the the empty text area would not be used and we'd rely on the object to return an empty text value later. This had an implicit coupling of the empty value. In b3ebb74 the component and 'complex' panel creation is done in helper classes. In the helper class we now conditionally create the entire panel where this text area would have been used, this avoids the double check for the empty 'additonal data' case.

**Comment by DanVanAtta** [comment]
Thank you for feedback @ssoloff ; Working on this I re-visited how easy it is to follow the overall flow, I thought there were some issues and tried to fix things by breaking down the concerns of component creation, layout into panels, and layout into the frame. With luck this makes it a lot easier to follow, let me know what you think.

File: game-core/src/main/java/games/strategy/debug/error/reporting/ErrorReportWindowModel.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+ */
+class ErrorReportWindowModel {
+
+  private final JFrame parent;
+
+  @Nullable
+  private final String attachedData;
+
+
+  ErrorReportWindowModel(final JFrame parent, final LogRecord logRecord) {
```
[the remark this answers is not in the document]
Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/debug/error/reporting/ErrorReportWindowModel.java]: In retrospect here, as you noticed, the constructors needed cleaning up so we would have just one, thank you for the suggestions :+1: 

Done in: 8a878f672

File: game-core/src/main/java/games/strategy/debug/error/reporting/ErrorReportWindowModel.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        logRecord.getSourceMethodName(),
+        Arrays.toString(logRecord.getThrown().getStackTrace()));
+  }
+
+  ErrorReportWindowModel(final JFrame parent, final String consoleText) {
+    this.parent = parent;
+    attachedData = consoleText;
+  }
+
+  ErrorReportWindowModel(final JFrame parent) {
```
[the remark this answers is not in the document]
Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/debug/error/reporting/ErrorReportWindowModel.java]: Done in: 8a878f672

File: game-core/src/main/java/games/strategy/debug/error/reporting/ErrorReportWindowModel.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            .toString())
+        .setVisible(true);
+  }
+
+  /**
+   * On submit we'll ask the user for confirmation if they'd like to send the error erport. If yes,
+   * then the error report will be uploaded.
+   */
+  Runnable submitAction(
+      final Component button,
```
[the remark this answers is not in the document]
Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/debug/error/reporting/ErrorReportWindowModel.java]: This one no, legacy refactoring: previously the submitAction needed a button reference to enable/disable.

Unused param removed in: 06226f3b4

File: game-core/src/main/java/games/strategy/debug/error/reporting/ErrorReportWindow.java [hunk 6]
```diff
@@ -1,82 +1,218 @@
...
-                .guiReader(guiReader)
-                .build()))
-            .addHorizontalStrut(100)
-            .add(errorReportComponents.newCancelButton(frame::dispose))
-            .addHorizontalStrut(30)
-            .build())
-        .build());
-    return frame;
+
+    final Panels panelBuilder = new Panels(frame, components);
```
[the remark this answers is not in the document]
Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/debug/error/reporting/ErrorReportWindow.java]: Good call-out: I had the same consideration and updated the class name but did not also get the variable renamed as well.

Variable renamed to `panels`:  6c4674fb2
